addons/mail/models/mail_render_mixin.py

```diff
@@ -251,11 +222,12 @@ def _prepend_preview(self, html, preview):
             preview = preview.strip()
 
         if preview:
+            # todo: check with tde if it is ok to remove the .format()
             html_preview = Markup("""
                 <div style="display:none;font-size:1px;height:0px;width:0px;opacity:0;">
-                   {}
+                   %s
                 </div>
-            """).format(preview)
+            """ % preview)
```

Comment:
Why this change ?

Reply:
In `action_send_mail` we call this method with `mailing._prepend_preview(mailing.body_html, mailing.preview),`
There is a test that will trigger those instructions (`test_mailing_w_blacklist_opt_out`).

If we consider that `mailing.preview` can contain qweb instructions (as it is the case in `test_mailing_w_blacklist_opt_out`), and if we format the preview with qweb (`<t t-out="...">`) the `<t>` tag will be escaped and the test fails.

Reply:
Ok. Indeed preview is something that can be dynamic.

Reply:
Ok but the security implication is different. It means the `preview` content must be trusted. It looks fine in the current code but worth keeping in mind.

Reply:
Ok, I am changing `mail.preview` to be an html field.

Reply:
That's typically a confusion between "HTML escaping" and "template language escaping", cfr my comment in `html_sanitize()`.

The `preview` field is supposed to contain TEXT, not HTML. See the [original implementation](https://github.com/odoo/odoo/commit/e39a87383244b7c0198ee120b20af60e4ae16432#diff-fff5506914deb026d9a83738d9f6bcb65a213194a2360d4b3891ddb7b56c654cR225) which had an explicit `html_escape()`.

The tooltip on the `preview` field does not state that expressions are allowed, but the implementation allowed it (by chance?) by inserting the preview value inside the `body` *before* the jinja rendering. 

So now we have a **conflict** between the format of the `preview` field (text with inline expressions) and the format of the `body` where it is injected (XML with QWeb).

If you want to keep the current feature you'll need to:

- convert the jinja-like inline expressions of `preview` into qweb expressions on-the-fly, while HTML-escaping anything else (😵)
- or change the `preview` field into a real HTML field with QWeb edition (😵 too)

Neither option looks good.

The simplest alternative would perhaps be to drop that feature and only allow static text inside `preview`. But perhaps @std-odoo can comment about that ;-)


PS: to clarify, the current code breaks the case where the user puts this perfectly legit static text inside `preview`: "Latest R&D news 🤓" (`&` needs escaping). Or this less-legit text: `<script>alert(document.cookie)</script>` (shouldn't execute code upon preview)

Reply:
It was not "by chance". It is wanted, even if tooltip does not explicitly tell it. If content is dynamic, no reason to make preview static.

Reply:
@Goaman No do not change "mailing.preview" to an HTML field, it's used as a preview when you open your email in Gmail (e.g.) and it make no sense to put HTML code in that field
![screenshot_spect](https://user-images.githubusercontent.com/54316226/123915049-5df04200-d980-11eb-8ab7-fee1b219e882.png)

And indeed, we wanted it to be dynamic and to be a "text" (not HTML code) so you should keep the `.format`.
What about just letting the flow use "Small Qweb" to render it ? Or maybe HTML "unescape" before "Small Qweb" rendering if the issue is that some python code might be escaped.
(in the end the problem come from the fact that we use 2 different template languages in the same template)

Reply:
@std-odoo thing is, there is no separate rendering step for `preview` at the moment. It's merged into the body [here](https://github.com/odoo/odoo/blob/master/addons/mass_mailing/models/mailing.py#L636), before it gets rendered. And later `mail_compose_message.render_message()` handles it. You'd need to add a separate rendering step for `preview` somewhere - including perhaps a separate support for it inside `mail.compose.message`?

Reply:
I did not think to render it and then add it in the body, but just to HTML escape it (as we do now) and in the "Small Qweb" method, unescape before rendering (so it will unescape all "Small Qweb" template).

So this part of code stay unchanged, the only change would be
```python
def parse_small_qweb(text):
    text = html_unescape(text)
    ...
```

Would it work ?

Reply:
I've made a commit that allow the rendering of the preview with the `inline_template`: https://github.com/odoo-dev/odoo/commit/7da80aa5552fa4e647c0d56e2744b8c69a11a267